contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler}
+import akka.stream._
+
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Pause/ Resume a Flow
+  */
+trait ValveSwitch {
+  /**
+    * Change the state of the valve
+    *
+    * @param mode expected mode to switch on
+    *
+    * @return
+    *         Future[true] if the mode change
+    *         Future[false] if the mode didn't change
+    */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+/**
+  * Valve permit to control a flow (pause/resume) by using a ValveSwitch.
+
+  * After calling [[ValveSwitch#flip(mode: SwitchMode)]] the running instance of the [[Graph]] of [[FlowShape]] that materialized to the
+  * [[Valve]] will change the state of the valve
+  *
+  * After calling [[ValveSwitch#flip(Close)]]: the valve state is changed to Close, and if any push is requested, the current element to send is save in the buffer
+  *
+  * After calling [[ValveSwitch#flip(Open)]]: the valve state is changed to Open, if a buffered Element is detected it will be automatically be push
+  *
+  * @param mode : state of the valve at the startup of the flow (by default Open)
+  */
+class Valve[A](mode: SwitchMode = SwitchMode.Open) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  override val shape = FlowShape(Inlet[A]("valve.in"), Outlet[A]("valve.out"))
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: SwitchMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    import shape._
+
+    var bufferedElement = Option.empty[A]
+
+    val switch = new ValveSwitch {
+
+      override def flip(_mode: SwitchMode): Future[Boolean] = {
+        val promise = Promise[Boolean]()
+
+        val callback = getAsyncCallback[Option[A]]{ element =>
+          val currentMode = mode
+          Try{
+            if (mode == SwitchMode.Close) {
+              mode = SwitchMode.Open
+              element.foreach(push(out, _))
+            }
+            else {
+              mode = SwitchMode.Close
+            }
+          } match {
+            case Success(_) => promise.success(currentMode != _mode)
+            case Failure(ex) => promise.failure(ex)
```

Comment:
the only case i can see where an exception can happen is in the ```push()``` logic (but i wasn t sure that could be happen).
Maybe the Try logic is a little bit too much